pkg/exchange/coinbase/exchage.go

```diff
@@ -0,0 +1,309 @@
+	types.Interval1m:  60,
+	types.Interval5m:  5 * 60,
+	types.Interval15m: 15 * 60,
+	types.Interval1h:  60 * 60,
+	types.Interval6h:  6 * 60 * 60,
+	types.Interval1d:  24 * 60 * 60,
+}
+
+func (e *Exchange) SupportedInterval() map[types.Interval]int {
+	return supportedIntervalMap
+}
+
+func (e *Exchange) IsSupportedInterval(interval types.Interval) bool {
+	_, ok := supportedIntervalMap[interval]
+	return ok
+}
+
+// ExchangeMinimal
+func (e *Exchange) Name() types.ExchangeName {
+	return types.ExchangeCoinBase
+}
+
+func (e *Exchange) PlatformFeeCurrency() string {
+	return PlatformToken
+}
+
+// ExchangeAccountService
+func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
+	if err := queryAccountLimiter.Wait(ctx); err != nil {
+		return nil, err
+	}
+
+	balances, err := e.QueryAccountBalances(ctx)
+	if err != nil {
+		return nil, err
+	}
+	account := types.NewAccount()
+	account.UpdateBalances(balances)
+	return account, nil
+}
+
+func (e *Exchange) QueryAccountBalances(ctx context.Context) (types.BalanceMap, error) {
+	if err := queryAccountLimiter.Wait(ctx); err != nil {
+		return nil, err
+	}
+
+	req := e.client.NewGetBalancesRequest()
+	accounts, err := req.Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+	balances := make(types.BalanceMap)
+	for _, cbBalance := range accounts {
+		cur := strings.ToUpper(cbBalance.Currency)
+		balances[cur] = *toGlobalBalance(cur, &cbBalance)
+	}
+	return balances, nil
+}
+
+// ExchangeTradeService
+func (e *Exchange) SubmitOrder(ctx context.Context, order types.SubmitOrder) (createdOrder *types.Order, err error) {
+	req := e.client.NewCreateOrderRequest().OrderType(string(order.Type)).Side(string(order.Side)).ProductID(toLocalSymbol(order.Symbol)).Size(order.Quantity.String())
+	res, err := req.Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &types.Order{
+		Exchange: types.ExchangeCoinBase,
+		UUID:     res.ID,
+	}, nil
+}
+
+func (e *Exchange) QueryOpenOrders(ctx context.Context, symbol string) ([]types.Order, error) {
+	cbOrders, err := e.queryOrdersByPagination(ctx, toLocalSymbol(symbol), []string{"open"})
+	if err != nil {
+		return nil, err
+	}
+	orders := make([]types.Order, 0, len(cbOrders))
+	for _, cbOrder := range cbOrders {
+		orders = append(orders, *toGlobalOrder(&cbOrder))
+	}
+	return orders, nil
+}
+
+func (e *Exchange) queryOrdersByPagination(ctx context.Context, symbol string, status []string) ([]api.Order, error) {
+	sortedBy := "created_at"
+	sorting := "desc"
+	before := time.Now()
+	localSymbol := toLocalSymbol(symbol)
+	paginationLimit := 1000
+	getOrdersReq := e.client.NewGetOrdersRequest()
+	getOrdersReq.ProductID(localSymbol).Status(status).SortedBy(sortedBy).Sorting(sorting).Before(before).Limit(paginationLimit)
+	cbOrders, err := getOrdersReq.Do(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get orders")
+	}
+
+	if len(cbOrders) < paginationLimit {
+		return cbOrders, nil
+	}
+
+	done := false
+	for {
+		if done {
```

Comment:
you need to handle the ctx.Done() in the loop.

Reply:
Done.